common/cgroup/cgroup.go

```diff
@@ -69,32 +76,68 @@ func MaybeAddIsolator(opts []*unit.UnitOption, isolator string, limit *resource.
 	if limit == nil {
 		return opts, nil
 	}
-	if IsIsolatorSupported(isolator) {
+	isSupported, err := IsIsolatorSupported(isolator)
+	if err != nil {
+		return nil, err
+	}
+
+	if isSupported {
 		opts, err = isolatorFuncs[isolator](opts, limit)
-		if err != nil {
-			return nil, err
-		}
 	} else {
 		fmt.Fprintf(os.Stderr, "warning: resource/%s isolator set but support disabled in the kernel, skipping\n", isolator)
 	}
+
+	if err != nil {
+		return nil, err
+	}
 	return opts, nil
 }
 
 // IsIsolatorSupported returns whether an isolator is supported in the kernel
-func IsIsolatorSupported(isolator string) bool {
-	if files, ok := cgroupControllerRWFiles[isolator]; ok {
+func IsIsolatorSupported(isolator string) (bool, error) {
+	isUnified, err := IsCgroupUnified("/")
+	if err != nil {
+		return false, errwrap.Wrap(errors.New("error determining cgroup version"), err)
+	}
+
+	if isUnified {
+		controllers, err := GetEnabledControllers()
+		if err != nil {
+			return false, errwrap.Wrap(errors.New("error determining enabled controllers"), err)
+		}
+		for _, c := range controllers {
+			if c == isolator {
+				return true, nil
+			}
+		}
+		return false, nil
+	}
+
+	if files, ok := legacyCgroupControllerRWFiles[isolator]; ok {
 		for _, f := range files {
 			isolatorPath := filepath.Join("/sys/fs/cgroup/", isolator, f)
 			if _, err := os.Stat(isolatorPath); os.IsNotExist(err) {
-				return false
+				return false, nil
 			}
 		}
-		return true
+		return true, nil
+	}
+	return false, nil
+}
+
+// IsCgroupUnified checks if cgroup mounted at /sys/fs/cgroup is
+// the new unified version (cgroup v2)
+func IsCgroupUnified(root string) (bool, error) {
```

Comment:
The way it work is the following:

1. If there's cgroup v2 in `/sys/fs/cgroup` we use that.

2. If there's cgroup v1 in `/sys/fs/cgroup` we do our mounting in the container directory and we use that.

3. If there's no cgroup on `/sys/fs/cgroup` we create it and mount it ourselves on the host, and we do the same as with cgroup v1.

4. If there's no cgroup on `/sys/fs/cgroup` and there's no support in the kernel, there's no way that machine can run containers.

So unless we want to distinguish the 4th case to inform the user (which has nothing to do with what's in `/sys/fs/cgroup`), I think there's only 2 valid outputs for this function so I'd leave it like it is now.